R/glossary.R

```diff
@@ -0,0 +1,390 @@
+#' @details
+#' For all entries in the new object, the abbreviation is set to the label.
+#' See [update_abbrev()] to customize the abbreviation.
+#'
+#' Entries that have duplicate labels will be dropped with a warning.
+#'
+#' @return
+#' A new glossary object.
+#'
+#' @seealso [read_glossary()], [update_abbrev()]
+#'
+#' @examples
+#'
+#' as_glossary(wt = "weight", ht = "height", bmi = "body mass index")
+#'
+#' as_glossary(wt = "weight", ht = "height", wt = "baseline weight")
+#'
+#' l <- list(VPC = "visual predictive check", tz = "timezone")
+#'
+#' as_glossary(l)
+#'
+#' as_glossary(l, wt = "weight")
+#'
+#' @seealso [update_abbrev()]
+#' @export
+as_glossary <- function(...) {
+  x <- purrr::list_flatten(list(...))
+  if(!is.list(x) || !is_named(x)) {
+    abort("items in `...` must resolve to a named list.")
+  }
+  if(anyDuplicated(names(x))) {
+    duplicates <- duplicated(names(x))
+    drop <- paste0(names(x)[duplicates], ": ", x[duplicates])
+    names(drop) <- rep("*", length(drop))
+    warn("Dropping duplicate glossary labels:", body = drop)
+    x <- x[!duplicates]
+  }
+  type_chr <- vapply(x, inherits, what = "character", FUN.VALUE = TRUE)
+  len <- vapply(x, length, FUN.VALUE = 1)
+  if(!all(type_chr & len==1)) {
+    abort("Every `glossary` entry must have type character and length 1.")
+  }
+  glossary <- Map(x, names(x), f = function(def, label) {
+    data <- c(definition = def, abbreviation = label)
+    as_glossary_entry(data)
+  })
+  class(glossary) <- c("tex_glossary", "glossary", "list")
+  glossary
+}
+
+as_glossary_entry <- function(x) {
+  x <- as.list(x)
+  class(x) <- c("glossary_entry", "list")
+  x
+}
+
+abort_bad_glo_labels <- function(x, what) {
+  if(!all(what %in% names(x))) {
+    bad <- setdiff(what, names(x))
+    names(bad) <- rep("x", length(bad))
+    msg <- c("Requested definitions not found in glossary object", bad)
+    abort(msg)
+  }
+}
+
+#' Return formatted notes from a tex glossary file or glossary object
+#'
+#' @param x path to a tex glossary file, a glossary object, or a list that can
+#' be coerced to a glossary object with [as_glossary()].
+#'
+#' @inheritParams read_glossary
+#' @inheritParams st_notes_glo
+#'
+#' @examples
+#' file <- system.file("glo", "glossary.tex", package = "pmtables")
+#'
+#' glossary_notes(file, WT, CRCL)
+#'
+#' g <- as_glossary(list(ss = "steady state", ALB = "albumin", WT = "weight"))
+#'
+#' glossary_notes(g, ALB, ss)
+#'
+#' @seealso [st_notes_glo()], [read_glossary()]
+#' @export
+glossary_notes <- function(x, ...) UseMethod("glossary_notes")
+
+#' @param ... unquoted glossary labels to be included in the notes text.
+#' @param format passed to [read_glossary()].
+#' @rdname glossary_notes
+#' @export
+glossary_notes.character <- function(x, ..., format = guess_glo_fmt(x)) {
+  x <- read_glossary(x, format = format)
+  glossary_notes(x, ...)
+}
+
+#' @rdname glossary_notes
+#' @export
+glossary_notes.list <- function(x, format = guess_glo_fmt(x), ...) {
```

Comment:
Is `format` a pasto from `glossary_notes.character`? Including `format` here will trigger an error:

```R
glossary_notes(list(ss = "steady state", ALB = "albumin", WT = "weight"))
#> Error in `guess_glo_fmt()` at pmtables/R/glossary.R:191:3:
#> ! Could not guess glossary format from file extension.
#> Run `rlang::last_trace()` to see where the error occurred.
#> Warning messages:
#> 1: In if (ext == "tex") return("tex") :
#>   the condition has length > 1 and only the first element will be used
#> 2: In if (ext %in% c("yml", "yaml")) return("yaml") :
#>   the condition has length > 1 and only the first element will be used
```

Reply:
I can't explain it either. Fixed this signature and the function; added tests for making glossary notes from a list. The original was hubris on my part. But I was on a roll ...